### PR TITLE
Fix StackOverflow errors for long recurrences

### LINKED ISSUE #49: StackOverflowError for long recurrences
Consider the following code

```julia
import Random
import Flux                                                                                                                                                                                                                                                                                                                                                
import Flux.Tracker                                                                                                                                                                                                                                                                                                                                        
                       
Random.seed!(2018)
                                                                                                                                                                                                                                                                                                                                    
N = (1 << 10)                                                                                                                                                                                                                                                                                                                                              
x = randn(N)                                                                                                                                                                                                                                                                                                                                               
                                                                                                                                                                                                                                                                                                                                                           
                                                                                                                                                                                                                                                                                                                                                           
function compute_ema(half_life, init_state)                                                                                                                                                                                                                                                                                                                
    alpha = log(2) / half_life                                                                                                                                                                                                                                                                                                                             
    beta = exp(-alpha)                                                                                                                                                                                                                                                                                                                                     
                                                                                                                                                                                                                                                                                                                                                           
    h = init_state                                                                                                                                                                                                                                                                                                                                         
    for i = 1:N                                                                                                                                                                                                                                                                                                                                            
        h = beta * h + alpha * x[i]                                                                                                                                                                                                                                                                                                                        
    end                                                                                                                                                                                                                                                                                                                                                    
    h                                                                                                                                                                                                                                                                                                                                                      
end                                                                                                                                                                                                                                                                                                                                                        
                                                                                                                                                                                                                                                                                                                                                           
grad = Tracker.gradient(compute_ema, 1000.0, 0.0)                                                                                                                                                                                                                                                                                                          
println(grad)  
```

This outputs
```
(-1.736128260716541e-5 (tracked), 0.4917510370131168 (tracked))
```

However, if I increase `N` to `(1 << 13)` then it has a `StackOverflowError`
```
ERROR: LoadError: StackOverflowError:
Stacktrace:
 [1] accum!(::Flux.Tracker.Grads, ::Flux.Tracker.Tracked{Float64}, ::Flux.Tracker.TrackedReal{Float64}) at /home/ermartin/.julia/packages/Flux/oN61x/src/tracker/back.jl:100
 [2] back(::Flux.Tracker.Grads, ::Flux.Tracker.Tracked{Float64}, ::Flux.Tracker.TrackedReal{Float64}) at /home/ermartin/.julia/packages/Flux/oN61x/src/tracker/back.jl:115
 [3] (::getfield(Flux.Tracker, Symbol("##4#5")){Flux.Tracker.Grads})(::Flux.Tracker.Tracked{Float64}, ::Flux.Tracker.TrackedReal{Float64}) at /home/ermartin/.julia/packages/Flux/oN61x/src/tracker/back.jl:106
 [4] foreach at ./abstractarray.jl:1836 [inlined]
 [5] back_(::Flux.Tracker.Grads, ::Flux.Tracker.Call{getfield(Flux.Tracker, Symbol("##250#253")){Flux.Tracker.TrackedReal{Float64},Flux.Tracker.TrackedReal{Float64}},Tuple{Flux.Tracker.Tracked{Float64},Flux.Tracker.Tracked{Float64}}}, ::Flux.Tracker.TrackedReal{Float64}) at /home/ermartin/.julia/packages/Flux/oN61x/src/tracker/back.jl:106
 [6] back(::Flux.Tracker.Grads, ::Flux.Tracker.Tracked{Float64}, ::Flux.Tracker.TrackedReal{Float64}) at /home/ermartin/.julia/packages/Flux/oN61x/src/tracker/back.jl:118
 ... (the last 4 lines are repeated 1 more time)
 [11] (::getfield(Flux.Tracker, Symbol("##4#5")){Flux.Tracker.Grads})(::Flux.Tracker.Tracked{Float64}, ::Flux.Tracker.TrackedReal{Float64}) at /home/ermartin/.julia/packages/Flux/oN61x/src/tracker/back.jl:106
 [12] back_(::Flux.Tracker.Grads, ::Flux.Tracker.Call{getfield(Flux.Tracker, Symbol("##250#253")){Flux.Tracker.TrackedReal{Float64},Flux.Tracker.TrackedReal{Float64}},Tuple{Flux.Tracker.Tracked{Float64},Flux.Tracker.Tracked{Float64}}}, ::Flux.Tracker.TrackedReal{Float64}) at ./abstractarray.jl:1836
 ... (the last 7 lines are repeated 6859 more times)
 [48026] back(::Flux.Tracker.Grads, ::Flux.Tracker.Tracked{Float64}, ::Int64) at /home/ermartin/.julia/packages/Flux/oN61x/src/tracker/back.jl:118
 [48027] FluxML/Flux.jl#4 at /home/ermartin/.julia/packages/Flux/oN61x/src/tracker/back.jl:106 [inlined]
 [48028] foreach at ./abstractarray.jl:1836 [inlined]
 [48029] back_(::Flux.Tracker.Grads, ::Flux.Tracker.Call{getfield(Flux.Tracker, Symbol("##226#229")),Tuple{Flux.Tracker.Tracked{Float64},Flux.Tracker.Tracked{Float64}}}, ::Int64) at /home/ermartin/.julia/packages/Flux/oN61x/src/tracker/back.jl:106
 [48030] back(::Flux.Tracker.Grads, ::Flux.Tracker.Tracked{Float64}, ::Int64) at /home/ermartin/.julia/packages/Flux/oN61x/src/tracker/back.jl:118
 [48031] FluxML/Flux.jl#6 at /home/ermartin/.julia/packages/Flux/oN61x/src/tracker/back.jl:131 [inlined]
 [48032] (::getfield(Flux.Tracker, Symbol("##9#11")){getfield(Flux.Tracker, Symbol("##6#7")){Flux.Tracker.Params,Flux.Tracker.TrackedReal{Float64}}})(::Int64) at /home/ermartin/.julia/packages/Flux/oN61x/src/tracker/back.jl:140
 [48033] gradient(::Function, ::Float64, ::Vararg{Float64,N} where N) at /home/ermartin/.julia/packages/Flux/oN61x/src/tracker/back.jl:152
 [48034] top-level scope at none:0
 [48035] include at ./boot.jl:317 [inlined]
 [48036] include_relative(::Module, ::String) at ./loading.jl:1044
 [48037] include(::Module, ::String) at ./sysimg.jl:29
 [48038] exec_options(::Base.JLOptions) at ./client.jl:231
in expression starting at /home/ermartin/test.jl:22
```

My guess is that this happens because of the recursive nature of the backwards algorithm. I don't know how possible it is to modify backprop to use an iterative algorithm instead of recursive. I also don't know if another autodiff system such as Zygote would handle this better.

Any suggestions for workarounds or plans to eventually fix this issue? Support for fast auto-diff over algorithms on long sequences would be valuable for time series analysis.

### PR DESCRIPTION
**Pull request by AStupidBear**
This is an improved version of Flux#iterate which fixes #49

### CODE REVIEW

**Comment by ChrisRackauckas** [comment]
@DhairyaLGandhi ?

**Comment by DhairyaLGandhi** [comment]
This effectively replaces the recursive function with an explicit nested loop.

The change to the `TrackedArray` struct itself could potentially break user code (o don't think many people rely on it so it can be fine), but it's not immediately obvious when the grads would be of a different type. Could you maybe add an example that that catches?

**Comment by AStupidBear** [comment]
When `data` is an `Ajoint`, the `grad` will be of type `Array`, therefore raising errors if we use a single type. We didn't encounter this before because the `grad` field was not used in the backward phase, but now the backward pipeline is changed.

**Comment by DhairyaLGandhi** [comment]
Genrrally lgtm , would need a check against Flux to see that gradients aren't affected.

File: src/back.jl [hunk 1]
```diff
@@ -14,51 +14,65 @@ init_grad(x) = zero(x)
...
-    Δ
-  end
-  if ref == 0
-    back_(x.f, grad, once)
-    once && !x.isleaf && (x.f = Call(missing, ()))
-  end
-  return
+    seen = Set{UInt64}(objectid(x))
+    if isdefined(x, :grad)
+      x.grad = zero_grad!(x.grad)
```
Comment (DhairyaLGandhi) [inline src/back.jl:70]: Why are we calling this here? Seems like we should get the right behaviour without this as well.

Comment (AStupidBear) [reply src/back.jl:70]: This is previously done in the `scan` https://github.com/FluxML/Tracker.jl/blob/b460dcf5199c7d64ec5ea9e2392fa131f7369b2e/src/back.jl#L19 but now `scan` is removed.

**Comment by DhairyaLGandhi** [comment]
Ah, missed that.

alright, I'll run a test suite against the model zoo of this tonight and if it checks out, we can merge this quickly

**Comment by ChrisRackauckas** [comment]
Bump?

**Comment by DhairyaLGandhi** [comment]
On it

**Comment by DhairyaLGandhi** [comment]
lgtm

**Pull request merged**